Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,448 @@
+/-
+Copyright (c) 2023 Mohanad Ahmed. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mohanad Ahmed
+-/
+
+import Mathlib.Data.Matrix.ColumnRowPartitioned
+import Mathlib.Data.Matrix.Rank
+import Mathlib.LinearAlgebra.Matrix.SVD.Reindex
+
+/-! # Singular Value Decomposition
+
+This file provides the SVD theorem which decomposes a real/complex matrix into left
+eigenvectors, singular values block diagonal matrix and right eigenvectors.
+
+Any matrix A (M × N) with rank r = A.rank and with elements in the field ℝ or ℂ can be decomposed
+into the product of three matrices `A = UξVᴴ` where:
+
+  * U: an `M × M` matrix containing the left eigenvectors of the matrix
+  * ξ: an `M × N` matrix with an `r × r` block in the upper left corner with nonzero singular values
+  * V: an `N × N` matrix containing the right eigenvectors of the matrix
```

Comment:
I am completely confused by this docstring. To me, singular values of a matrix `A` have to do with the successive maximum expansion rates of the matrix, so I don't see how left or right eigenvectors of the matrix `A` would appear in a singular value decomposition.

Reply:
1. The singular values for me are square roots of eigenvalues. They are also what remains of the matrix transformation if we rotate the source and target vector spaces until the transformation is diagonal.

2. The Singular Value Decomposition is a generalization of the Eigendecompositon. We sometimes abuse terminology and call the left singular vectors left eigenvectors and similarly for the right singular vectors.

If we write left singular vector in place of left eigenvectors and right singular vector in place of right eigenvectors, does that work for you?

```
* U: an `M × M` matrix containing the left singular vector of the matrix
* ξ: an `M × N` matrix with an `r × r` block in the upper left corner with nonzero singular values
* V: an `N × N` matrix containing the right singular vectorof the matrix
 ```

3. I did not define a function for singular values. I only defined the matrix of singular values which has the same dimensions as the original matrix A. It is also partitioned such that it is upper block of size $r \times r$ contains the $r$ non-zero singular values.

I have no Idea what the maximum expansion rate of a matrix is.